File: practice/convert/example2_book_library/book-library/src/main/java/dev/starryeye/book_library/application/user/query/result/GetUsersResult.java
package dev.starryeye.book_library.application.user.query.result;

import dev.starryeye.book_library.domain.user.User;

import java.util.List;

public record GetUsersResult(
        List<GetUserResult> users
) {

    public static GetUsersResult of(List<User> users) {
        return new GetUsersResult(
                users.stream()
                        .map(user -> new GetUserResult(
                                user.getId(),
                                user.getUsername(),
                                user.getAge()
                        ))
                        .toList()
        );
    }
}

File: practice/convert/example2_book_library/kotlin/book-library/src/main/kotlin/dev/starryeye/book_library/application/user/query/result/GetUsersResult.kt
package dev.starryeye.book_library.application.user.query.result

import dev.starryeye.book_library.domain.user.User

data class GetUsersResult(
    val users: List<GetUserResult>,
) {

    companion object {
        fun of(users: List<User>): GetUsersResult {
            return GetUsersResult(
                users.map { user ->
                    GetUserResult(
                        id = user.id!!, // DB 거쳐서 나온 엔티티에는 id 가 항상 존재한다.
                        username = user.username,
                        age = user.age
                    )
                }
            )
        }
    }
}